alf/algorithms/rlpd_algorithm.py

```diff
@@ -0,0 +1,213 @@
+            env=env,
+            config=config,
+            critic_loss_ctor=critic_loss_ctor,
+            target_entropy=target_entropy,
+            prior_actor_ctor=prior_actor_ctor,
+            target_kld_per_dim=target_kld_per_dim,
+            initial_log_alpha=initial_log_alpha,
+            max_log_alpha=max_log_alpha,
+            target_update_tau=target_update_tau,
+            target_update_period=target_update_period,
+            parameter_reset_period=parameter_reset_period,
+            dqda_clipping=dqda_clipping,
+            actor_optimizer=actor_optimizer,
+            critic_optimizer=critic_optimizer,
+            alpha_optimizer=alpha_optimizer,
+            checkpoint=checkpoint,
+            debug_summaries=debug_summaries,
+            name=name)
+
+        assert self._act_type == ActionType.Continuous, (
+            "RLPD algorithm only supports continuous action spaces.")
+
+        self._num_critic_targets = num_critic_targets
+
+    def _compute_critics(self,
+                         critic_net,
+                         observation,
+                         action,
+                         critics_state,
+                         replica_consensus='mean',
+                         sample_subset=False,
+                         apply_reward_weights=True):
+        observation = (observation, action)
+        # critics shape [B, replicas]
+        critics, critics_state = critic_net(observation, state=critics_state)
+
+        # For multi-dim reward, do
+        # [B, replicas * reward_dim] -> [B, replicas, reward_dim]
+        # For scalar reward, do nothing
+        if self.has_multidim_reward():
+            remaining_shape = critics.shape[2:]
+            critics = critics.reshape(-1, self._num_critic_replicas,
+                                      *self._reward_spec.shape,
+                                      *remaining_shape)
+            if self._act_type == ActionType.Discrete:
+                # permute: [B, replicas, reward_dim, num_actions]
+                #       -> [B, replicas, num_actions, reward_dim]
+                order = [0, 1, -1] + list(
+                    range(2, 2 + len(self._reward_spec.shape)))
+                critics = critics.permute(*order)
+
+        if sample_subset:
+            critics = critics[:,
+                              torch.randperm(self._num_critic_replicas
+                                             )[:self._num_critic_targets], ...]
+
+        if replica_consensus == 'min':
+            if self.has_multidim_reward():
+                sign = self.reward_weights.sign()
+                critics = (critics * sign).min(dim=1)[0] * sign
+            else:
+                critics = critics.min(dim=1)[0]
+        elif replica_consensus == 'mean':
+            if self.has_multidim_reward():
+                sign = self.reward_weights.sign()
+                critics = (critics * sign).mean(dim=1) * sign
+            else:
```

Comment:
This branch is unnecessary for 'mean'

Reply:
Also curious the motivation of the ``mean`` branch? Any benefits?

Reply:
> This branch is unnecessary for 'mean'

I feel that it is needed in order to get the mean critic values across all critics for actor training?

Reply:
> Also curious the motivation of the `mean` branch? Any benefits?

This is the default setting of RLPD to get a concensus q_value for actor training. Given that RLPD might use more than two critics, taking a `min` as in SAC would be too conservative. Moreover, from my previous experience, as long as critics are trained with conservative target critic values, there is no need to be conservative in actor training.

Reply:
I see. Thanks!

Reply:
Oh I see, thanks, updated.

Reply:
I had to revert to keep this branch since `self.reward_weights` by default is `None` for scalar reward.

Reply:
Why can't you change the whole branch as the following?
```python
elif replica_consensus == 'mean':
    critics = critics.mean(dim=1)
```

Reply:
Oh yes, that should work well.